### ISSUE TITLE
watch's glob seems to work differently than `ls | glob`

### ISSUE THREAD

**Issue by colemickens** [nushell/nushell#7022]
### Describe the bug

```
ls | glob '<[!.]*/>[!.]*'
```

will recursively list files but ignore dot-files/dirs

but

```
▶ watch . -v --glob=<[!.]*/>[!.]* { |op, path, new_path| $"($op) ($path) ($new_path)"}
Absolute glob path: "/home/cole/code/nixcfg/<[!.]*/>[!.]*"
Now watching files at "/home/cole/code/nixcfg". Press ctrl+c to abort.
NoticeWrite("/home/cole/code/nixcfg/foo")
Write("/home/cole/code/nixcfg/foo")
Matches glob: false
Create("/home/cole/code/nixcfg/.git/index")
Matches glob: false
```

just results in no matches

### How to reproduce

Try to use `watch` without triggering on `.git` changes.

### Expected behavior

There to be some way to achieve this with the built in glob feature

### Configuration

```
| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.70.0                                          |
| branch             |                                                 |
| commit_hash        |                                                 |
| build_os           | linux-x86_64                                    |
| build_target       | x86_64-unknown-linux-gnu                        |
| rust_version       | rustc 1.64.0                                    |
| cargo_version      | cargo 1.64.0                                    |
| pkg_version        | 0.70.0                                          |
| build_time         | 1980-01-01 00:00:00 +00:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  |        |
```

**Comment by colemickens** [comment]
Oh I guess `glob` is its own command

**Comment by rgwood** [comment]
Hmm. I wrote `watch` so this is probably my fault. That’s a tricky glob pattern!

I’ll take a look tomorrow.

**Comment by sholderbach** [comment]
I think this points to a larger issue we have: we have our own glob implementation in `nu-glob` used in the filesystem commands (`ls` etc. and in this case `watch`)  and also have the `glob` command for which we import the globbing behavior from the `wax` crate. 
I think in the longer term we should only have one source of truth for the globbing.

**Comment by colemickens** [comment]
globbing doesn't really respect the ordering the same way bash does either.

"foo/{z,y,x}" will match "foo/x" "foo/y" "foo/z"

which in my case, break things and means I have to use a different construct

**Comment by fdncred** [comment]
> I think in the longer term we should only have one source of truth for the globbing.

Agreed @sholderbach. The reason I added `wax` and the `glob` command for globbing was because we had to strip some functionality from `nu-glob` to get it to work right with `ls`.

**Comment by rgwood** [comment]
So I guess the conclusion here is that `watch` should switch to `wax` for globbing?

**Comment by fdncred** [comment]
@rgwood another option is to allow `watch` to take an `input` from the `glob` command. e.g.
```
glob 'some convoluted glob syntax' | watch . { |op, path, new_path| $"($op) ($path) ($new_path)"}
```

**Comment by rgwood** [comment]
Hmm, that's a clever idea but I don't think it would work for `watch`. 

`glob ... | watch ...` would gather a list of files and then tell `watch` to watch only those files. But `watch` needs to be able to report on files that are created _after_ `watch` is launched.

**Comment by melMass** [comment]
What about making watch support multiple arguments? for instance a subcommand `watch files`

```nu
watch files **/*.rs {|i| print $i}
```

**Comment by rgwood** [comment]
@melMass

```bash
watch files **/*.rs {|i| print $i}
```
I think this is the same as what you can do right now with ` watch . --glob=**/*.rs { |i| print $i }`. Am I missing something?

```bash
watch files (<command that returns a list>) {|i| print $i}
```
Is `<command that returns a list>` returning a list of glob patterns? That could work but it feels a bit too complicated to me.

**Comment by rgwood** [comment]
^ we discussed this on Discord: https://discord.com/channels/601130461678272522/615329862395101194/1051209969439101010

I think we came to the conclusion that the current `watch <path> --glob=<glob> <closure>` design is OK (but `watch` should handle more sophisticated glob patterns).

**Comment by instantepiphany** [comment]
> ^ we discussed this on Discord: https://discord.com/channels/601130461678272522/615329862395101194/1051209969439101010
> 
> I think we came to the conclusion that the current `watch <path> --glob=<glob> <closure>` design is OK (but `watch` should handle more sophisticated glob patterns).

I've just encountered this. I want to run `cargo run --bin server` to build a specific binary in a rust workspace whenever the files in `server` or `lib` (local dependency) change. 
I want this to happen for `.rs` and `.toml` files. 

There is another workspace folder, `client` which should not be watched as it is not a dependency of `server`. 

So with the current glob limitations in `watch`, I:
1. Can't specify that only files ending in `.rs` or `.toml` should be watched (ie with `**/*.{rs,toml}`)
2. Can't specify that only files in `server` or `lib` should be watched (ie with `**/{server,lib}/**`)
3. Can't ignore specific folders (ie like `glob` can via `--exclude`). 


If `watch` was updated to use `wax`, or if `watch` supported multiple globs, this would be possible (via `wax` "alternatives" as in my examples, or by supplying multiple globs like `**/server/*.rs`, `**/server/Cargo.toml`, `**/lib/*.rs`, `**/lib/Cargo.toml`).

I suppose I could do another layer of filtering inside the closure, but that feels inefficient and awkward.

**Comment by fdncred** [comment]
The `glob` command also doesn't accept multiple separate globs on the command line. Although, I'd like it to. Seems like we could do it with the `any` functionality of wax but I could never get it integrated properly. The `glob` command does offer a completely different syntax though. It is sometimes helpful to be able to do things like `glob **/*.{rs,toml} --depth 2`.

**Comment by instantepiphany** [comment]
> The `glob` command also doesn't accept multiple separate globs on the command line. Although, I'd like it to. Seems like we could do it with the `any` functionality of wax but I could never get it integrated properly. The `glob` command does offer a completely different syntax though. It is sometimes helpful to be able to do things like `glob **/*.{rs,toml} --depth 2`.

Thats the alternative (`alternatives`, ha) that I mentioned in my comment.

If `wax` was used as the library for `nu-glob` then I could have used `watch . --glob **/{server,lib}/**/*.{rs,toml} {|| cargo run --bin server}` and that would have worked fine. 

Also unfortunate is that this glob matches 0 files, but doesn't error. I was scratching my head for quite a while before I realised that `glob` behaves differently to `nu-glob` used by `watch` - my first response was to put that glob pattern into `glob <pattern>`  so was confused when it produced exactly the files I wanted it to 😄

**Comment by fdncred** [comment]
> ... I was scratching my head for quite a while before I realised that glob behaves differently to nu-glob ...

`nu-glob` was meant to be a more simplistic globbing crate, whereas `glob` using `wax` was meant to be more "advanced". The two aren't necessarily compatible with each other.

I'm not sure if anyone is against switching glob crates in the watch command. What may be better is removing the globbing from `watch` and passing it in so one could do `glob some_complicated_globbing | watch --param1 --param2`. Then it would be composable and more nushell-y. If you wanted `nu-glob` functionality you could do `ls some_nu_glob | get name | watch --param1 --parm2`.

**Comment by instantepiphany** [comment]
> > ... I was scratching my head for quite a while before I realised that glob behaves differently to nu-glob ...
> 
> `nu-glob` was meant to be a more simplistic globbing crate, whereas `glob` using `wax` was meant to be more "advanced". The two aren't necessarily compatible with each other.

This makes sense - but for someone new to `nu`, it is a fair assumption that "globbing" in `nu` is consistent. Silently using different globbing logic will absolutely confuse users.


> I'm not sure if anyone is against switching glob crates in the watch command. What may be better is removing the globbing from `watch` and passing it in so one could do `glob some_complicated_globbing | watch --param1 --param2`. Then it would be composable and more nushell-y. If you wanted `nu-glob` functionality you could do `ls some_nu_glob | get name | watch --param1 --parm2`.

I thought about this when I encountered the problem. 

I think the trick with passing the glob paths in to `watch` as piped input is that if the command that `watch` runs generates files that would and should match the filter, they won't get picked up unless the user restarts watch. 

Whereas if `watch` is passed an argument, it can reexecute the glob after each time the requested command is invoked by `watch`, and get a new, current list of paths.

**Comment by fdncred** [comment]
> I thought about this when I encountered the problem.

makes sense
